brayns/network/entrypoints/InspectEntrypoint.cpp

```diff
@@ -21,10 +21,121 @@
 
 #include "InspectEntrypoint.h"
 
+namespace
+{
+class SceneInspector
+{
+public:
+    SceneInspector(brayns::Engine &engine)
+        : _engine(engine)
+    {
+    }
+
+    brayns::InspectResult inspect(const brayns::Vector2f &position)
+    {
+        auto pickResult = _pickOSPRayScene(position);
+        if (!pickResult.hasHit)
+        {
+            brayns::InspectResult result;
+            result.hit = false;
+            return result;
+        }
+
+        auto &instance = _findHittedModel(pickResult.instance);
+        const auto inspectContext = _buildInspectContext(pickResult);
+        auto metadata = brayns::JsonObject();
+
+        auto &model = instance.getModel();
+        model.onInspect(inspectContext, metadata);
+
+        _releaseOSPRayHandles(pickResult);
+
+        return _buildResult(inspectContext, instance, metadata);
+    }
+
+private:
+    OSPPickResult _pickOSPRayScene(const brayns::Vector2f &position)
+    {
+        auto &frameBuffer = _engine.getFrameBuffer();
+        auto &renderer = _engine.getRenderer();
+        auto &camera = _engine.getCamera();
+        auto &scene = _engine.getScene();
+
+        auto frameBufferHandle = frameBuffer.handle();
+        auto rendererHandle = renderer.handle();
+        auto cameraHandle = camera.handle();
+        auto sceneHandle = scene.handle();
+
+        auto x = position.x;
+        auto y = position.y;
+
+        OSPPickResult result;
+        ospPick(&result, frameBufferHandle, rendererHandle, cameraHandle, sceneHandle, x, y);
+        return result;
+    }
+
+    brayns::ModelInstance &_findHittedModel(OSPInstance hittedInstance)
+    {
+        auto &scene = _engine.getScene();
+        auto &models = scene.getModels();
+        auto &instances = models.getAllModelInstances();
+
+        auto begin = instances.begin();
+        auto end = instances.end();
+        auto instanceIterator = std::find_if(
+            begin,
+            end,
+            [&](brayns::ModelInstance *instancePtr) { return instancePtr->handle() == hittedInstance; });
+
+        // Shouldn't happen, but..
+        if (instanceIterator == end)
+        {
+            throw brayns::InternalErrorException("Could not find the hitted model instance in the scene");
+        }
+
+        return **instanceIterator;
+    }
+
+    brayns::InspectContext _buildInspectContext(OSPPickResult &osprayPickResult)
+    {
+        const auto &ospHitPosition = osprayPickResult.worldPosition;
+        auto hitPosition = brayns::Vector3f(ospHitPosition[0], ospHitPosition[1], ospHitPosition[2]);
+        auto geometricModel = osprayPickResult.model;
+        auto primitiveIndex = osprayPickResult.primID;
+
+        return brayns::InspectContext{hitPosition, geometricModel, primitiveIndex};
+    }
+
+    brayns::InspectResult _buildResult(
+        const brayns::InspectContext &context,
+        const brayns::ModelInstance &instance,
+        brayns::JsonObject &metadata)
```

Comment:
const brayns::JsonObject & ?

Here making a move would be better (pass by copy and then move) but I think Poco::JSON::Object doesn't implement move construction / assignment (to be checked).